docs/src/pages/guides/composition/composition.md

```diff
@@ -127,40 +127,70 @@ Here is a demo with [React Router DOM](https://github.com/ReactTraining/react-ro
+by using `ReactDOM.findDOMNode`. This function is deprecated in favor of `ref` and
+ref forwarding. However, only the following component types can be given a `ref`:
 
-- class components
-- ref forwarding components (`React.forwardRef`)
-- built-in components e.g. `div` or `a`
+* Any Material-UI component
+* class components i.e. `React.Component` or `React.PureComponent`
+* built-in (or host) components e.g. `div` or `button`
+* [React.forwardRef components](https://reactjs.org/docs/react-api.html#reactforwardref)
+* [React.lazy components](https://reactjs.org/docs/react-api.html#reactlazy)
+* [React.memo components](https://reactjs.org/docs/react-api.html#reactmemo)
 
-If this is not the case we will issue a prop type warning similar to:
+If you don't use one of the above types in Material-UI components you might see a warning from
+React in your console similar to 
+> Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
+
+Be aware that you still get this warnings for lazy and memo components if their
+wrapped component can't hold a ref.
+
+In some instances we issue an additional warning to help debugging similar to 
 > Invalid prop `component` supplied to `ComponentName`. Expected an element type that can hold a ref.
 
-In addition React will issue a warning.
 
-You can fix this warning by using `React.forwardRef`. Learn more about it in
-[this section in the official React docs](https://reactjs.org/docs/forwarding-refs.html).
+We will only cover the two most common use cases. For more information see [this section in the official React docs](https://reactjs.org/docs/forwarding-refs.html).
+
+```diff
+- const MyButton = props => <div role="button" {...props} />
++ const MyButton = React.forwardRef((props, ref) => <div role="button" {...props} ref={ref} />)
+<Button component={MyButton} />
+```
+
+```diff
+- const SomeContent = props => <div {...props}>Hello, World!</div>
++ const SomeContent = React.forwardRef((props, ref) => <div {...props} ref={ref}>Hello, World!</div>)
+<Tooltip title="Hello, again.">
+```
 
 To find out if the Material-UI component you're using has this requirement, check
 out the the props API documentation for that component. If you need to forward refs
 the description will link to this section.
 
 ### Caveat with StrictMode or unstable_ConcurrentMode
 
-If you pass class components to the `component` prop and don't run in strict mode you won't have to change anything
-since we can safely use `ReactDOM.findDOMNode`. For function components, however, you have
-to wrap your component in `React.forwardRef`:
+If you use class components for the cases described above you will still see
+warnings in `React.StrictMode` and `React.unstable_ConcurrentMode`. We use
+`ReactDOM.findDOMNode` internally for backwards compatibility. You can use `React.forwardRef`
+and a designated prop in your class component to forward the `ref` to a built-in
```

Comment:
I'll be honest I don't know how to properly explain this. Maybe I'm using the term `ref` wrong maybe it isn't well defined.

If I refer to the term `ref` then I mean the return value from e.g. `React.createRef()`. So the ref is the full object `{ current: instance }`. What we end up is attaching the ref object `{ current: ... }` to the host component i.e. `<div ref={{ current: ... }} />`. Or in other words the ref is the thing you pass to the `ref` prop<sup>1</sup> of a component. You'll probably still catch me doing `<Component ref={ref => ...} />` where naming the parameter `ref` is probably a misnomer. `<Component ref={current => ...} />` should be clearer.

The DOM element will then be `ref.current`. Or more generally `ref.current` will point to the component instance. For host components rendered with `react-dom` it will be an instance of a DOM `Element`.

Please tell me if that doesn't make sense. In the end the most important thing is that we don't confuse the reader. It's not that important to be technically accurate but that one understands how to fix the issue at hand.

<sup>1</sup> `ref` gets a special treatment so calling it a ref might not be 100% accurate.

Reply:
```suggestion
and a designated prop in your class component to forward the `ref` to a DOM
```

Reply:
What I found confusing was the terminology "built-in component". What exactly does that refer to? 

(I also overlooked the fact that a ref is a pointer to the full object, not just a discrete DOM element, so my suggested replacement wasn't any more helpful.)